content/doc/developer/security/csp.adoc

```diff
@@ -0,0 +1,110 @@
+Look up the elements by their ID or class or similar characteristics, then call `#addEventListener` on them.
+Be mindful of Jenkins's extensibility, so consider including plugin names in element class names or IDs to prevent unintentional conflicts with other plugins.
+
+Use `Behaviour#specify` to add event handlers to elements that may be dynamically added to the page, for example as part of AJAX responses.
+One common instance of this is in configuration forms: `renderOnDemand` is used by common form elements like `hetero-list` to load parts of the page only as the form is being changed.
+The code that adds content from AJAX responses dynamically to the page needs to call `Behaviour#applySubtree` on the newly added content.
+
+Examples of this are: https://github.com/jenkinsci/jenkins/pull/5514[jenkinsci/jenkins#5514]
+
+=== Legacy JavaScript `checkUrl` validation
+
+Form validation using manually specified `checkUrl` parameters currently supports inline JS for its "legacy" mode.
+It looks like the following:
+
+[source, html]
+<f:textbox checkUrl="'${rootURL}/${h.jsStringEscape(it.url)}checkText?value='+encodeURIComponent(this.value)+'" … />
+
+This combines inline JS and building parts of the string using JEXL expressions in Jelly, with different ways to escape different parts of the content to prevent injection vulnerabilities.
+
+Instead, use the _modern_ `checkUrl` mode, which as of Jenkins 2.360 requires the `checkDependsOn` attribute to be set (but it can be an empty string).
+This mode will automatically add the current form elements value as the query parameter called `value`, so the above examples can be simplified to the following:
+
+[source, html]
+<f:textbox checkUrl="${rootURL}/${it.url}checkText" checkDependsOn="" … />
+
+Examples of this are: https://github.com/jenkinsci/jenkins/pull/6856[jenkinsci/jenkins#6856] https://github.com/jenkinsci/jenkins/pull/6857[jenkinsci/jenkins#6857]
+
+To pass additional values, specify the respective form field names as part of the `checkDependsOn` string.
+
+If you need to pass parameters that are not represented as form fields, the following options exist as of Jenkins 2.360:
+
+* Define a new form validation endpoint.
+  This can be a viable option when it's a boolean value (2 endpoints instead of one).
+* Define a hidden form field (wrap it in `f:invisibleEntry`) with the expected `name` and `value` and specify it in `checkDependsOn`.
+  Make sure to ignore it otherwise.
+  See https://github.com/jenkinsci/jenkins/pull/6859[jenkinsci/jenkins#6859] for an example.
+
+
+=== `eval` calls
+
+`eval` should not be used to interpret a string as JS code.
+
+Depending on your use case, different solutions are possible.
+
+To parse JSON, use `JSON.parse` instead.
+See https://github.com/jenkinsci/jenkins/pull/6868[jenkinsci/jenkins#6868] for an example.
+
+To invoke a callback, have the caller define a global function and pass its name as an argument.
```

Comment:
Having to define globals seems awkward. Do you have an example of this case?

Reply:
`var foo = …;` (or `let foo = …;`) outside any block makes it global scoped.

Reply:
Yeah, I just meant that the general scenario of when you would need to do this is not really clear to me.

Reply:
It's the part of https://github.com/jenkinsci/jenkins/pull/6865 related to `renderOnDemand.jelly`.